lib/visitor.js

```diff
@@ -0,0 +1,61 @@
+"use strict";
+
+const hasOwn = Object.prototype.hasOwnProperty;
+
+// This Visitor API was inspired by a similar API provided by ast-types:
+// https://github.com/benjamn/ast-types/blob/master/lib/path-visitor.js
+
+function Visitor() {
+  this.visit = this.visit.bind(this);
+  this.visitWithoutReset = this.visitWithoutReset.bind(this);
+  this.traverse = this.traverse.bind(this);
+}
+
+module.exports = Visitor;
+const Vp = Visitor.prototype;
+
+Vp.visit = function (path) {
+  this.reset.apply(this, arguments);
+  this.visitWithoutReset(path);
+};
+
+Vp.visitWithoutReset = function (path) {
+  const value = path.getValue();
+  if (Array.isArray(value)) {
+    path.each(this.visitWithoutReset);
+  } else if (path.getNode() === value) {
+    const method = this["visit" + value.type];
+    if (typeof method === "function") {
+      // Note: the method must call this.traverse(path) to continue
+      // traversing.
+      method.call(this, path);
+    } else {
+      this.traverse(path);
+    }
+  }
+};
+
+Vp.traverse = function (path) {
+  if (! path.valueIsNode()) {
+    return;
+  }
+
+  const node = path.getValue();
+
+  Object.keys(node).forEach(function (key) {
+    if (key === "loc") {
+      return;
+    }
+
+    if (key.charAt(0) === "_") {
+      return;
+    }
+
+    const value = node[key];
+    if (! (value && typeof value === "object")) {
+      return;
+    }
```

Comment:
With `ast-types`, we know exactly what fields a given AST node is supposed to have, based on its `.type` (see [here](https://github.com/benjamn/ast-types/blob/e855515735dd84e415dd03319e0cd6e1ecb59204/lib/path-visitor.js#L223)). Because we don't know anything about the expected structure of AST nodes, there's a chance here that we're visiting properties of nodes that don't matter. For example, the `.loc.{start,end}` object isn't going to matter for this transform, and Babel attaches "private" properties (containing metadata, not AST nodes) by prefixing their names with `_`.

My general point is just that any intelligent short cuts we can take to ignore whole subtrees of the AST could pay off big time. Like, if we knew whether or not a subtree contained any `import` or `export` tokens, that might allow us to skip lots of subtrees. That kind of thing.